Tutorial
========

A quick tutorial to use CodAn.

Download the file ["transcripts.fa"](https://github.com/pedronachtigall/CodAn/blob/master/tutorial/transcripts.fa), which has 500 full transcript sequences from *Danio rerio*.

Download the ["VERT_full.zip"](https://github.com/pedronachtigall/CodAn/blob/master/models/VERT_full.zip) and ["VERT_partial.zip"](https://github.com/pedronachtigall/CodAn/blob/master/models/VERT_partial.zip) models available at the subfolder ["CodAn/models/"](https://github.com/pedronachtigall/CodAn/tree/master/models).

Running the Full model
======================

Decompress the model "VERT_full.zip":
```
unzip VERT_full.zip
```

Then, run CodAn:
```
codan.py -t transcripts.fa -m VERT_full
```

Expected print message at the terminal:
```
  _____           _  ___        
 /  __ \         | |/ _ \       
 | /  \/ ___   __| / /_\ \_ __  
 | |    / _ \ / _` |  _  | '_ \ 
 | \__/\ (_) | (_| | | | | | | |
  \____/\___/ \__,_\_| |_/_| |_|
                          
        
2019-10-25 09:23:13 >>>> starting CodAn (v1.0 September 2019)...
	transcript file -> transcripts.fa
	model -> VERT_full
	strand prediction -> both
	number of threads -> 1
2019-10-25 09:23:13 >>>> CDS prediction...
2019-10-25 09:24:42 >>>> retrieving sequences...
	number of transcripts -> 500
	number of predictions -> 487
		predictions at plus strand -> 486
		predictions at minus strand -> 1
2019-10-25 09:24:42 >>>> prediction finished!
	CDS -> CodAn_output/ORF_sequences.fasta
	3'UTR -> CodAn_output/3utr_sequences.fasta
	5'UTR -> CodAn_output/5utr_sequences.fasta
	GTF file -> CodAn_output/annotation.gtf
```


The output files will be available at the "CodAn_output" folder. If you want to specify the name of the output folder use the option ```-o output_folder```. CodAn uses only one thread by default ```-c 1```, but the user can specify the use of more threads by using the option ```-c int``` to accelerate the process.

By default, CodAn performs the predition on both strands of the input transcripts (```-s both```). But the user can specify the strand to be used in the prediction using the option ```-s str```. To run the predictions only in one of the strands, use the parameter ```-s plus``` for the plus strand and the parameter ```-s minus``` for the minus strand.

Running the Partial model
=========================

Decompress the model "VERT_partial.zip":
```
unzip VERT_partial.zip
```

Then, run CodAn:
```
codan.py -t transcripts.fa -m VERT_partial -o CodAn_partial_test
```

Expected print message at the terminal:
```
  _____           _  ___        
 /  __ \         | |/ _ \       
 | /  \/ ___   __| / /_\ \_ __  
 | |    / _ \ / _` |  _  | '_ \ 
 | \__/\ (_) | (_| | | | | | | |
  \____/\___/ \__,_\_| |_/_| |_|
                          
        
2019-10-25 09:32:17 >>>> starting CodAn (v1.0 September 2019)...
	transcript file -> transcripts.fa
	model -> VERT_partial
	strand prediction -> both
	number of threads -> 1
2019-10-25 09:32:17 >>>> CDS prediction...
2019-10-25 09:33:43 >>>> retrieving sequences...
	number of transcripts -> 500
	number of predictions -> 488
		predictions at plus strand -> 487
		predictions at minus strand -> 1
2019-10-25 09:33:43 >>>> prediction finished!
	CDS -> CodAn_output/ORF_sequences.fasta
	3'UTR -> CodAn_output/3utr_sequences.fasta
	5'UTR -> CodAn_output/5utr_sequences.fasta
	GTF file -> CodAn_output/annotation.gtf
```

Running the BLAST search
========================
To run the BLAST search in the coding sequences predicted jut use the option ```-b``` and indicate the path to a protein database pre-mounted, such as swissprot (ftp://ftp.ncbi.nlm.nih.gov/blast/db/), or generate your own protein database with the ```makeblastdb``` function from [NCBI-BLAST](https://www.ncbi.nlm.nih.gov/books/NBK279671/).
```
makeblastdb -dbtype prot -in protein_sequences.fa -out blastDB
```

Then, run CodAn with the following options:
```
codan.py -t transcripts.fa -m VERT_full -o CodAn_test_blast -b blastDB
```

Outputs
=======
The basic output files generated by CodAn are three [FASTA](https://en.wikipedia.org/wiki/FASTA_format) files and one [GTF](https://www.ensembl.org/info/website/upload/gff.html) file.
The three FASTA files present the sequences of the CDS, 3'UTR and 5'UTR predcited by CodAn, whereas the GTF file has the annotations of the CDS, start and stop codon predicted in the transcripts.

If you use the BLAST search option, the output of the BLAST search will be generated in a subfolder named "blast_results", which will contain three FASTA files with the annotation of the best hit in the header and a ```report.log``` file. The ```report.log``` file is a tab-separated file with two columns: (1) the header of the transcript in the input file; and (2) the name of the best hit in the BLAST search.
